### ISSUE TITLE
Why is `optimizer.step()` called at mini-batch level instead of step level in PPO Actor?

### ISSUE THREAD

**Issue by zhaojingqian** [volcengine/verl#2793]
In `verl/workers/actor/dp_actor.py`, the optimizer update happens at the **mini-batch level** within the `update_policy` method:

```python
def update_policy(self, data: DataProto):
    metrics = {}
    for epoch in range(self.config.ppo_epochs):           # PPO epochs
        for batch_idx, data in enumerate(dataloader):     # Mini batches
            self.actor_optimizer.zero_grad()              # ← Clear gradients per mini-batch
            
            for micro_batch_idx, data in enumerate(micro_batches):  # Micro batches
                loss.backward()                            # Accumulate gradients
            
            grad_norm = self._optimizer_step()            # ← Update per mini-batch ❌
```

## Expected Behavior

From a conceptual standpoint, one **training step** should correspond to one **optimizer update**. Looking at the main training loop in `ray_trainer.py`:

```python
for epoch in range(self.config.trainer.total_epochs):
    for batch_dict in self.train_dataloader:              # ← This is one training step
        # 1. Rollout generation
        gen_batch_output = self.actor_rollout_wg.generate_sequences(gen_batch)
        
        # 2. Reward & advantage computation
        batch = compute_advantage(...)
        
        # 3. Policy update (calls update_policy)
        actor_output = self.actor_rollout_wg.update_actor(batch)
        
        # 4. Step increment
        self.global_steps += 1                            # ← One complete step
```

The entire `update_policy` call should be treated as **one step**, not multiple steps.

## Comparison with SFT Trainer

The SFT trainer in `verl/trainer/fsdp_sft_trainer.py` follows the correct pattern:

```python
def training_step(self, batch: TensorDict):
    self.optimizer.zero_grad()                            # ← Clear once per step
    
    micro_batches = batch.split(self.config.data.micro_batch_size_per_gpu)
    for micro_batch in micro_batches:
        loss = self._compute_loss_and_backward(batch=micro_batch) / n_micro_batches
    
    self.optimizer.step()                                 # ← Update once per step ✅
```

## Proposed Fix

The PPO actor should follow a similar pattern:

```python
def update_policy(self, data: DataProto):
    metrics = {}
    
    # Clear gradients once at the beginning of the step
    self.actor_optimizer.zero_grad()
    
    total_mini_batches = 0
    for epoch in range(self.config.ppo_epochs):
        for batch_idx, data in enumerate(dataloader):
            total_mini_batches += 1
            
            for micro_batch_idx, data in enumerate(micro_batches):
                # Only accumulate gradients, don't update parameters
                loss = policy_loss / (total_mini_batches * self.gradient_accumulation)
                loss.backward()
                
                # Collect metrics
                append_to_dict(metrics, data)
    
    # Update optimizer once after processing the entire rollout batch
    grad_norm = self._optimizer_step()
    data = {"actor/grad_norm": grad_norm.detach().item()}
    append_to_dict(metrics, data)
    
    return metrics
```

## Questions

1. **Is the current mini-batch level updating intentional?** If so, what's the reasoning behind it?

2. **Does this affect PPO algorithm correctness?** Multiple optimizer steps per training step might change the learning dynamics compared to standard PPO.

3. **Should we normalize gradients differently?** Currently, gradients are normalized by `gradient_accumulation`, but with multiple mini-batches, should we also account for the number of mini-batches?

**Comment by Lvleix** [comment]
this is standard PPO I think🤔

**Comment by SamComber** [comment]
Yeah historically PPO implementations divide the global batch into minibatches for sequential parameter updates, it goes back to the Schulman paper I believe https://arxiv.org/pdf/1707.06347 (section 5. algorithm)

**Comment by zhaojingqian** [comment]
> Yeah historically PPO implementations divide the global batch into minibatches for sequential parameter updates, it goes back to the Schulman paper I believe https://arxiv.org/pdf/1707.06347 (section 5. algorithm)

Thanks for the answer!
